### PR TITLE
azure: support external file dependencies

### LINKED ISSUE #3540: azure: support external dependencies and outputs
As the title says, we currently do not support external dependencies and outputs from azure. So, 
`dvc import-url` and friends don't work.

#### Implementation
It should be trivial to implement.  `dvc` uses etag to ensure that it's same file. So, just need to implement following methods in [dvc/remote/azure.py](https://github.com/iterative/dvc/blob/master/dvc/remote/azure.py) similar to one in [dvc/remote/s3.py](https://github.com/iterative/dvc/blob/master/dvc/remote/s3.py):

https://github.com/iterative/dvc/blob/a1fe6c6f44777463876ad24ee0d162173999f9d3/dvc/remote/s3.py#L78-L79

> From Ruslan's reply on Discord: https://discordapp.com/channels/485586884165107732/485596304961962003/691327678271193160

Then:
- Implement a `DependencyAzure` and `OutputAzure` classes. Refer [DependencyS3](https://github.com/iterative/dvc/blob/master/dvc/dependency/s3.py) and [OutputS3](https://github.com/iterative/dvc/blob/master/dvc/output/s3.py) for the example.

- Add `OutputAzure` on [OUTS](https://github.com/iterative/dvc/blob/02f92b3ac85532d24ae94727a6071376dbf35452/dvc/output/__init__.py#L16), [OUTS_MAP](https://github.com/iterative/dvc/blob/02f92b3ac85532d24ae94727a6071376dbf35452/dvc/output/__init__.py#L24) and [CHECKSUMS_SCHEMA](https://github.com/iterative/dvc/blob/02f92b3ac85532d24ae94727a6071376dbf35452/dvc/output/__init__.py#L46) on [dvc/output/__init__.py](https://github.com/iterative/dvc/blob/master/dvc/output/__init__.py)

- Also `DependencyAzure` needs to be added on [DEPS](https://github.com/iterative/dvc/blob/02f92b3ac85532d24ae94727a6071376dbf35452/dvc/dependency/__init__.py#L19) and [DEPS_MAP](https://github.com/iterative/dvc/blob/02f92b3ac85532d24ae94727a6071376dbf35452/dvc/dependency/__init__.py#L29) on [dvc/dependency/__init__.py](https://github.com/iterative/dvc/blob/master/dvc/dependency/__init__.py).

P.S. I could be missing a few things though. :(

### PR DESCRIPTION
**Pull request by steffansluis**
Documention PR: https://github.com/iterative/dvc.org/pull/1337

- Implement get_file_checksum with get_etag
- Implement AzureDependency
- Add AzureDependency to DEPS and DEPS_MAP
- Add unit tests for AzureDependency

Fixes #3540.

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here. If the CLI API is changed, I have updated [tab completion scripts](https://github.com/iterative/dvc/tree/master/scripts/completion).

* [x] ❌ I will check DeepSource, CodeClimate, and other sanity checks below. (We consider them recommendatory and don't expect everything to be addressed. Please fix things that actually improve code or fix bugs.)

### CODE REVIEW

**Comment by steffansluis** [comment]
I'm not really sure why the tests are failing on Travis. Locally, they pass having followed the instructions at [https://dvc.org/doc/user-guide/contributing/core#testing-remotes](https://dvc.org/doc/user-guide/contributing/core#testing-remotes).

File: tests/unit/dependency/test_azure.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from dvc.dependency.azure import AzureDependency
+from tests.unit.dependency.test_local import TestLocalDependency
+
+
+class TestAzureDependency(TestLocalDependency):
```
Comment (skshetry) [inline tests/unit/dependency/test_azure.py:5]: @steffansluis, Are you sure it's running locally?

`AzureRemote` remote parses path via regex:
https://github.com/iterative/dvc/blob/865d355554b27be5ea49640c14556c46f7906327/dvc/remote/azure.py#L22-L27

And, on `TestLocalDependency`, the path is of the following type which fails on regex parsing:
https://github.com/iterative/dvc/blob/f5055940a2f806ccdf8be0193115641effd69416/tests/unit/dependency/test_local.py#L14

Same thing is happening on `TestAzureOutput`. Maybe, change the path that's getting passed to `AzureRemote` for tests?

Comment (steffansluis) [reply tests/unit/dependency/test_azure.py:5]: (Having done `source tests/remotes_env`).

`python -m tests tests/unit/output/test_azure.py tests/unit/dependency/test_azure.py` shows (among other things) the following output:
```
[gw1] [ 25%] PASSED tests/unit/output/test_azure.py::TestLocalOutput::test_save_missing                                                                                                       
[gw3] [ 50%] PASSED tests/unit/dependency/test_azure.py::TestLocalDependency::test_save_missing                                                                                               
[gw0] [ 75%] PASSED tests/unit/output/test_azure.py::TestAzureOutput::test_save_missing                                                                                                       
[gw2] [100%] PASSED tests/unit/dependency/test_azure.py::TestAzureDependency::test_save_missing                                                                                               
```

The tests are basic copies of the tests for S3: https://github.com/iterative/dvc/blob/f5055940a2f806ccdf8be0193115641effd69416/tests/unit/dependency/test_s3.py#L5-L7

And looking at the tests for the HDFS output (which [also uses regexes](https://github.com/iterative/dvc/blob/f5055940a2f806ccdf8be0193115641effd69416/dvc/remote/hdfs.py#L22)) this should be fine?:
https://github.com/iterative/dvc/blob/f5055940a2f806ccdf8be0193115641effd69416/tests/unit/dependency/test_hdfs.py#L5-L7

File: dvc/remote/azure.py [hunk 2]
```diff
@@ -81,6 +81,15 @@ def blob_service(self):
             blob_service.create_container(self.path_info.bucket)
         return blob_service
 
+    def get_content_md5(self, path_info):
+        content_md5 = self.blob_service.get_blob_properties(
+            path_info.bucket, path_info.path
+        ).properties.content_settings.content_md5
```
Comment (efiop) [inline dvc/remote/azure.py:87]: Does it stay true for multipart objects?

Comment (efiop) [reply dvc/remote/azure.py:87]: Also, does the user set it or is it automatically set by azure?

Comment (efiop) [reply dvc/remote/azure.py:87]: We need very good tests for this to get it accepted, as otherwise it might be too dangerous to use.

Comment (steffansluis) [reply dvc/remote/azure.py]: For potential future reference: `Content-MD5` appears to not behave as specified in the [documentation of the REST API](https://docs.microsoft.com/en-us/rest/api/storageservices/put-blob). Namely, large files (I tested with a 4GB file) do not get assigned a `Content-MD5` property automatically. Hence, I've changed it back to use the etag instead

(see also: https://discordapp.com/channels/485586884165107732/565699007037571084/713462619469643796)

File: dvc/remote/azure.py [hunk 3]
```diff
@@ -27,6 +27,7 @@ def __init__(self, repo, config):
 
         url = config.get("url", "azure://")
         self.path_info = self.path_cls(url)
+        print(self.path_info)
```
Comment (efiop) [inline dvc/remote/azure.py:30]:
```suggestion
```

**Comment by efiop** [comment]
@steffansluis Btw, could you please update the docs in https://dvc.org/doc/command-reference/import-url ?

File: dvc/remote/azure.py (hunk 3 above)
Comment (steffansluis) [reply dvc/remote/azure.py:30]: Yeah will get rid of that :see_no_evil:

**Comment by steffansluis** [comment]
@efiop It's very basic but I think it should do: https://github.com/iterative/dvc.org/pull/1337

File: dvc/dependency/azure.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+from dvc.dependency.base import BaseDependency
+from dvc.remote.azure import AzureRemote
+
+
+class AzureDependency(BaseDependency):
```
Comment (efiop) [inline dvc/dependency/azure.py:5]:
```suggestion
class AzureDependency(BaseDependency, BaseOutput):
```

**Pull request merged**